ta/pkcs11/src/processing.c

```diff
@@ -747,3 +747,165 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
 
 	return PKCS11_CKR_OK;
 }
+
+enum pkcs11_rc entry_wrap_key(struct pkcs11_client *client,
+			      uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_attribute_head *proc_params = NULL;
+	struct pkcs11_object *wrapping_key = NULL;
+	struct pkcs11_object *key = NULL;
+	struct obj_attrs *req_attrs = NULL;
+	uint32_t wrapping_key_handle = 0;
+	uint32_t key_handle = 0;
+	uint32_t size = 0;
+	void *key_data = NULL;
+	size_t key_sz = 0;
+	void *out_buf = params[2].memref.buffer;
+	uint32_t out_size = params[2].memref.size;
+	enum processing_func function = PKCS11_FUNCTION_WRAP;
+
+	if (!client || ptypes != exp_pt ||
+	    (out_size && !out_buf))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &wrapping_key_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &key_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_one_attribute(&ctrlargs, &proc_params);
+	if (rc)
+		goto out;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	/* Return error if processing already active */
+	rc = get_ready_session(session);
+	if (rc)
+		goto out;
+
+	/* Check wrapping key handle */
+	wrapping_key = pkcs11_handle2object(wrapping_key_handle, session);
+	if (!wrapping_key) {
+		rc = PKCS11_CKR_WRAPPING_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	/* Check key handle */
+	key = pkcs11_handle2object(key_handle, session);
+	if (!key) {
+		rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	/* Check if mechanism can be used for wrapping function */
+	rc = check_mechanism_against_processing(session, proc_params->id,
+						function,
+						PKCS11_FUNC_STEP_INIT);
+	if (rc)
+		goto out;
+
+	/* Set the processing state to active */
+	rc = set_processing_state(session, function, wrapping_key, NULL);
```

Comment:
wrapping unwrapping are one shot functions (1single stage processing) hence no need to set and release the active processing. Call to `get_ready_session(session)` at line 803 ensures there is no pending active processing and that's enough.

Reply:
Ok

Reply:
Even though , we don't need to set or release active processing here, but since we later use init_symmteric_operation() which uses tee_operation etc maintained in session processing state. Hence, call to this function would be required. I realized that after I started getting crashes on removing the call to this function.